tests/main/upgrade-from-release/task.yaml

```diff
@@ -0,0 +1,76 @@
+summary: Ensure upgrades from release version of snap works
+
+systems: [ubuntu-1*-64, ubuntu-2*-64]
+
+prepare: |
+    #shellcheck source=tests/lib/pkgdb.sh
+    . "$TESTSLIB/pkgdb.sh"
+    distro_purge_package snapd
+
+restore: |
+    #shellcheck source=tests/lib/pkgdb.sh
+    . "$TESTSLIB/pkgdb.sh"
+    distro_install_build_snapd
+
+execute: |
+    #shellcheck source=tests/lib/systemd.sh
+    . "$TESTSLIB"/systemd.sh
+
+    . /etc/os-release
+    # trusty has no UBUNTU_CODENAME in /etc/os-release and we need to cheat
+    # because snapd was not part of the original 14.04 release so we get
+    # the latest version in security before it went ESM
+    if [ "${VERSION_ID}" = "14.04" ]; then
+        UBUNTU_CODENAME=trusty-security
+    fi
+
+    # on anything but 16.04 get the snapd from the current distro
+    # release (without -updates)
+    if [ "${VERSION_ID}" != "16.04" ]; then
+        echo "deb http://archive.ubuntu.com/ubuntu ${UBUNTU_CODENAME} main" > /tmp/no-updates.list
+        apt update -o dir::etc::sourcelist=/tmp/no-updates.list -o dir::etc::sourceparts=/tmp/not-exists
+        apt install -o dir::etc::sourcelist=/tmp/no-updates.list -o dir::etc::sourceparts=/tmp/not-exists -y snapd
+        apt update -qq
+    else
+        # 16.04 is ESM so get the latest version from the official archive
+        # (we can't get the very first version because it it's so old it
+        #  cannot run our test snaps)
+        apt install -y snapd
+    fi
+    case ${VERSION_ID} in
+       21.10)
+         apt list --installed snapd | MATCH '2.53\+21.10ubuntu1'
+         ;;
+       20.04)
+         apt list --installed snapd | MATCH '2.44.3\+20.04'
+         ;;
+       18.04)
+         apt list --installed snapd | MATCH '2.32.5\+18.04'
+         ;;
+       16.04)
+         apt list --installed snapd | MATCH '2.48.3'
+         ;;
+       14.04)
+         apt list --installed snapd | MATCH '2.37.4~14.04.1'
+         ;;
+    esac
+
+    echo "install a test service and a test command"
+    snap install go-example-webserver
+    wait_for_service snap.go-example-webserver.webserver.service
+    snap install test-snapd-tools
+    test-snapd-tools.echo hello | MATCH hello
+
+    echo "upgrade to current snapd"
+    if [ "${VERSION_ID}" = "14.04" ]; then
+        dpkg -i "$GOHOME"/snapd*.deb
+    else
+        apt install -y "$GOHOME"/snapd*.deb
+    fi
+
+    echo "snapd listens to requests"
+    snap list
```

Comment:
Would this fail with the unfixed version? As far as I understand, the check that we need to perform is that *during* the update we should have snapd available after the old package has been removed. So I was wondering if we could somehow simulate an upgrade of snapd along with another deb package generated by the test, and make sure (how? - I don't know) that the upgrade works like this:

1. old snapd is removed
2. our test package is removed (or installed, it doesn't matter -- what metters is that this step occurs between steps 1 and 3)
3. the new snapd is installed

Then in a debian script (prerm, postrm, postinstall or whatever works) we check that at step 2 we can call `snap list` without errors.

Reply:
This would fail with https://github.com/snapcore/snapd/pull/11680 on 16.04 - this test is specifically to ensure that snapd is still active after a deb restart. The check for the new `dh_systemd_start --upgrade-after-restart` is done in the new  `deb-restart-behavior` test that was added in PR#11680.

Reply:
Thanks for the explanation!